### DISCUSSION TITLE
HTS/HCS/HSC/HFS Interoperability

### DISCUSSION THREAD

**Discussion by andrewantar** [hiero-ledger/hiero-improvement-proposals#69] [Ideas]
I wanted to start a discussion of interoperability between Hedera services. This could become several HIPs since there are several “bridges” to build but some use cases come to mind based on discussions with developers:

- Smart contracts being able to mint/transfer HTS tokens
- HCS being able to mint/transfer tokens
- HCS storing things on File service
- HTS NFT’s being linked to URL on File service
- Smart contracts talking to HCS and file service
- HTS tokens linked to ownership defined in HCS topic

I’m sure there are more use cases including using any number of services concurrently for example there could be a DAO on HSC (smart contracts) that records events in HCS and also mints governance tokens on HTS in accordance with the rules/logic of the smart contract. HCS entries from HSC can also link to a HFS (file service) to store a script or URLs to a data dump. I actually have a real world use case like this I’m helping someone map out using Hedera. 

Another prime use case is minting NFTs on HTS and having the NFT represent ownership of a real world asset that is recorded in HCS. One example is using HTS to mint an NFT for a sneaker or tshirt whose provenance “birth certificate” is tagged on HCS using Atma.io. A physical NFC chip label that is linked to HCS can be applied to a physical product and that product can be sold or traded in a marketplace using its unique HTS NFT linked to its HCS NFC label.

**Comment by donaldthibeau** [comment]
Beyond mint and transfer, what other use cases would smart contracts be desired to provide for HTS tokens?

**Reply by bugbytesinc** [reply]
If keys exist to perform burn, confiscate, kyc, freeze etc....it would seem that for consistency a contract should be able to perform said actions - not sure how you reconcile key management, probably a topic for further discussion.

**Comment by bugbytesinc** [comment]
I might suggest baby steps:

1. Allow arbitrary transfers to piggy-back on any transaction (so a contract/account could be co-funded with an initial balance, or an HCS message could also transfer tokens/hbar)
2. Revisit potential expansion of scheduled transaction white list
3. Then, focus on adding interop to the HVM, assuming the community does not come up with something better in the form of a side-chain, but one feature at a time, say start with HTS transfers, HTS admin functions, followed by writing to HCS.....

**Comment by andrewantar** [comment]
There could be a smart contract protocol that governs data input into file
service and/or HCS and validators who vote in a time-boxed set of
rounds time-stamped on HCS to accept the new data into file service and
then the smart contract releases a HTS token from escrow once successfully
completed. This is a potential workflow for a variety of use cases that
demonstrates the need for interoperability between HCS, HTS, smart contract
service, and file service.
